Modules/CTP/src/CTPTrendingTask.cxx

```diff
@@ -0,0 +1,223 @@
+#include <QualityControl/DatabaseInterface.h>
+#include <QualityControl/MonitorObject.h>
+#include <QualityControl/RootClassFactory.h>
+#include <QualityControl/QcInfoLogger.h>
+#include <QualityControl/ActivityHelpers.h>
+#include <QualityControl/RepoPathUtils.h>
+#include <QualityControl/Reductor.h>
+#include <QualityControl/UserCodeInterface.h>
+#include <CCDB/BasicCCDBManager.h>
+#include <DataFormatsCTP/Configuration.h>
+#include <DataFormatsCTP/RunManager.h>
+
+#include <TCanvas.h>
+#include <TH1.h>
+
+using namespace o2::quality_control;
+using namespace o2::quality_control::core;
+using namespace o2::quality_control::postprocessing;
+using namespace o2::quality_control::repository;
+using namespace o2::quality_control_modules::ctp;
+using namespace o2::ctp;
+
+void CTPTrendingTask::configure(const boost::property_tree::ptree& config)
+{
+  mConfig = TrendingConfigCTP(getID(), config);
+}
+
+void CTPTrendingTask::initialize(Trigger, framework::ServiceRegistryRef services)
+{
+  // Preparing data structure of TTree
+  for (const auto& source : mConfig.dataSources) {
+    mReductors.emplace(source.name, root_class_factory::create<TH1ctpReductor>(source.moduleName, source.reductorName));
+  }
+  mTrend = std::make_unique<TTree>();
+  mTrend->SetName(PostProcessingInterface::getName().c_str());
+  mTrend->Branch("runNumber", &mMetaData.runNumber);
+  mTrend->Branch("time", &mTime);
+  for (const auto& [sourceName, reductor] : mReductors) {
+    mTrend->Branch(sourceName.c_str(), reductor->getBranchAddress(), reductor->getBranchLeafList());
+  }
+
+  getObjectsManager()->startPublishing(mTrend.get());
+}
+
+void CTPTrendingTask::update(Trigger t, framework::ServiceRegistryRef services)
+{
+  auto& qcdb = services.get<repository::DatabaseInterface>();
+
+  trendValues(t, qcdb);
+  generatePlots();
+}
+
+void CTPTrendingTask::finalize(Trigger t, framework::ServiceRegistryRef)
+{
+  generatePlots();
+}
+
+void CTPTrendingTask::trendValues(const Trigger& t, repository::DatabaseInterface& qcdb)
+{
+  mTime = activity_helpers::isLegacyValidity(t.activity.mValidity)
+            ? t.timestamp / 1000
+            : t.activity.mValidity.getMax() / 1000; // ROOT expects seconds since epoch.
+  mMetaData.runNumber = t.activity.mId;
+
+  map<string, string> metadata; // can be empty
+
+  std::string run = std::to_string(t.activity.mId);
+  CTPRunManager::setCCDBHost("https://alice-ccdb.cern.ch");
+  mCTPconfig = CTPRunManager::getConfigFromCCDB(t.timestamp, run);
+
+  bool foundInputs = false;
+
+  for (auto& dataSource : mConfig.dataSources) {
+    auto mo = qcdb.retrieveMO(dataSource.path, dataSource.name, t.timestamp, t.activity);
+    if (!mo) {
+      ILOG(Info, Support) << "no MO object" << ENDM;
+      continue;
+    }
+    ILOG(Info, Support) << "Got MO " << mo << ENDM;
+    TObject* obj = mo ? mo->getObject() : nullptr;
+    if (obj) {
+      foundInputs = true;
+      mReductors[dataSource.name]->update(obj);
+    }
+  }
+
+  if (!foundInputs) {
```

Comment:
I am wondering if you really should add another row in the tree if you manage to retrieve only some objects from the QCDB. What may happen is that your new row will consists of partially obsolete values, which might be misleading. Up to you to decide if this is what you want.

If not, you could invert the logic and have `bool inputMissing = false;` which is set to true when one of the `retrieveMO` calls fails, then you can skip `mTrend->Fill()`.